src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -1015,6 +1015,35 @@ void ShenandoahHeapRegion::promote_in_place() {
   ShenandoahYoungGeneration* young_gen = heap->young_generation();
   size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
 
+  assert(get_top_before_promote() == tams, "Cannot promote regions in place if top has advanced beyond TAMS");
+
+  // Since this region may have served previously as OLD, it may hold obsolete object range info.
```

Comment:
May be modify the comment to state something like:
// Clear any obsolete object range information; we'll register the live objects in the loop below,
// as well as registering and stamping free block fillers over dead objects.
// We also mark the entire range of the card table dirty; it'll be updated when we scan it next time
// for young pointers.

Reply:
We don't clear object registrations when an old region is freed because often times, the free region may next be put into service as a young-region, where the object registrations are "don't care" values.  I'll make the adjustment to the comment.

Reply:
In any run, the number of clearings is basically the number old region births if cleared at promotion or allocation of an old generation, or deaths if cleared at freeing of an old generation. In a non-terminating run,  the rate of births equals rate of deaths. So the difference is just _when_ we clear, not how much. The amount of clearing work in the long run is no different in either case. So the difference is cache pollution, mutator latency, and code maintenance. I'm guessing there's really no difference, since it all happens on concurrent GC threads. So code maintenance would probably trump everything, and thinking some more about it, I am guessing there's not much difference in that either in when the clearing is done.